### PR TITLE
OCE-25: Update situations with latest copy of alarm

### PR DESCRIPTION
**Pull request by smith-opennms**
This PR changes the behaviour of the Cluster Engine to update the Alarms in the Situation.
This behaviour change also fixes OCE-21 where the issue where the Diagnostic Text of the Situation was incorrect as this was due to the alarms not being updated and having the correct time.

* JIRA: https://issues.opennms.org/browse/OCE-25
* JIRA: https://issues.opennms.org/browse/OCE-21

### CODE REVIEW

File: datasource/common/src/main/java/org/opennms/oce/datasource/common/SituationBean.java [hunk 1]
```diff
@@ -45,7 +47,7 @@
 
     private Severity severity = Severity.INDETERMINATE;
     private List<ResourceKey> resourceKeys = new ArrayList<>();
-    private Set<Alarm> alarms = new LinkedHashSet<>();
+    private Map<String, Alarm> alarms = new HashMap<>();
```
Comment (mattixtech) [inline datasource/common/src/main/java/org/opennms/oce/datasource/common/SituationBean.java]: We could use a `LinkedHashMap` here if we want to preserve the ordering that the original implementation did. Not sure if we care.

Could also make this reference final as with my suggestion below we will only ever be modifying its contents.

File: datasource/common/src/main/java/org/opennms/oce/datasource/common/SituationBean.java [hunk 2]
```diff
@@ -97,15 +99,15 @@ public void addResourceKey(ResourceKey resourceKey) {
 
     @Override
     public Set<Alarm> getAlarms() {
-        return alarms;
+        return alarms.values().stream().collect(Collectors.toSet());
     }
 
     public void setAlarms(Set<Alarm> alarms) {
-        this.alarms = alarms;
+        this.alarms = alarms.stream().collect(Collectors.toMap(Alarm::getId, a -> a));
```
Comment (mattixtech) [inline datasource/common/src/main/java/org/opennms/oce/datasource/common/SituationBean.java]: To continue using the sorted map implementation:

```suggestion
        this.alarms.clear();
        this.alarms.putAll(alarms.stream().collect(Collectors.toMap(Alarm::getId, a -> a)));
```

File: engine/cluster/src/main/java/org/opennms/oce/engine/cluster/ClusterEngine.java [hunk 3]
```diff
@@ -352,6 +353,11 @@ public synchronized void onTick(long timestampInMillis) {
                     .flatMap(e -> e.getValue().stream())
                     .collect(Collectors.toList());
 
+            // refresh the situations with the existing alarms
+            candidateAlarms.stream().forEach(alarm -> {
```
Comment (mattixtech) [inline engine/cluster/src/main/java/org/opennms/oce/engine/cluster/ClusterEngine.java]:
```suggestion
            candidateAlarms.forEach(alarm -> {
```

**Comment by mattixtech** [comment]
Guess I was too slow!

**Comment by j-white** [comment]
We can still address the comments retroactively - needed to get these changes out into the wild.

**Comment by mattixtech** [comment]
I'll likely touch this class in the near future when refactor from bean to builder so I can take a look at it then.

**Pull request merged**